In [1]:
# Librerías básicas
import re


In [9]:
# STEP 1: Load document
from google.colab import files
uploaded = files.upload()

Saving 243591-DO-AVO-02-V01-251008.docx to 243591-DO-AVO-02-V01-251008.docx


In [10]:
!pip install python-docx


In [39]:
# Install docx2txt if not already installed
!pip install docx2txt

import docx2txt

# Leer TODO el documento
text = docx2txt.process("243591-DO-AVO-02-V01-251008.docx")

# Verificar tamaño
print("Longitud:", len(text))

# Mostrar TODO el texto (sin cortes)
print(text)

Longitud: 7434
ACTA DE VISITA DE OBRA Nº 002

243591-DO-AVO-002-V01_251008

		



MEJORA DE LA ACCESIBILIDAD DE LA ESTACIÓN DE SANTA PERPÈTUA DE LA MOGODA – LA FLORIDA, BARCELONA



 Fecha

Lugar

Hora de inicio

Hora de Finalización

08/10/25

Estación de Santa perpètua de la Mogoda - La Florida.

10:45

13:00



Principales asistentes:

Nombre

Empresa

Cargo

E-mail

Teléfono

Firma opcional









































































FUNCIÓN:  DC (director de contrato) DO (dirección de obra) DEO (dirección de ejecución) UTE (contratista)   

CSS (coordinador de seguridad)



OBJETO

Visita Semanal de obra



DOCUMENTACIÓN ENTREGADA O A ENTREGAR 









DOCUMENTACIÓN DE CONSULTA



Proyecto Constructivo de Mejora de la accesibilidad de la estación de Santa Perpetua de la Mogoda – La Florida, Barcelona.






ASUNTOS TRATADOS, ACUERDOS ALCANZADOS, COMUNICACIONES, INSTRUCCIONES:

Nº

Fecha

DESCRIPCIÓN DEL ASUNTO:

1.



AVANCE DE LA OBRA:

1.1. DO-DEO V

In [40]:
# STEP 2: Clean text
def clean_text(text):
    # Replace multiple whitespace characters (spaces, tabs, newlines) with a single space
    cleaned_text = re.sub(r'\s+', ' ', text)
    # Strip leading/trailing whitespace from the entire string
    cleaned_text = cleaned_text.strip()
    return cleaned_text

In [41]:
cleaned_document_text = clean_text(text)
print("Longitud del texto limpio:", len(cleaned_document_text))
print(cleaned_document_text)

Longitud del texto limpio: 7100
ACTA DE VISITA DE OBRA Nº 002 243591-DO-AVO-002-V01_251008 MEJORA DE LA ACCESIBILIDAD DE LA ESTACIÓN DE SANTA PERPÈTUA DE LA MOGODA – LA FLORIDA, BARCELONA Fecha Lugar Hora de inicio Hora de Finalización 08/10/25 Estación de Santa perpètua de la Mogoda - La Florida. 10:45 13:00 Principales asistentes: Nombre Empresa Cargo E-mail Teléfono Firma opcional FUNCIÓN: DC (director de contrato) DO (dirección de obra) DEO (dirección de ejecución) UTE (contratista) CSS (coordinador de seguridad) OBJETO Visita Semanal de obra DOCUMENTACIÓN ENTREGADA O A ENTREGAR DOCUMENTACIÓN DE CONSULTA Proyecto Constructivo de Mejora de la accesibilidad de la estación de Santa Perpetua de la Mogoda – La Florida, Barcelona. ASUNTOS TRATADOS, ACUERDOS ALCANZADOS, COMUNICACIONES, INSTRUCCIONES: Nº Fecha DESCRIPCIÓN DEL ASUNTO: 1. AVANCE DE LA OBRA: 1.1. DO-DEO Visita semanal de obra 08/10/25 Con fecha de 08 de octubre de 2025 las partes se reúnen en el emplazamiento de obra situado 

In [42]:
# STEP 3: Chunking
def chunk_text(text, chunk_size=500, overlap=50):
    chunks = []
    if not text:
        return chunks

    start = 0
    while start < len(text):
        end = start + chunk_size
        chunk = text[start:end]
        chunks.append(chunk)
        # Move the start for the next chunk, considering overlap
        start += chunk_size - overlap
        # Ensure start does not go past the end of the text if the last chunk was very small
        if start >= len(text) and len(chunk) < chunk_size:
            break
    return chunks

In [43]:
chunks = chunk_text(cleaned_document_text)
print(f"Número total de chunks: {len(chunks)}")
for i, chunk in enumerate(chunks):
    print(f"\n--- Chunk {i+1} ---")
    print(chunk)

Número total de chunks: 16

--- Chunk 1 ---
ACTA DE VISITA DE OBRA Nº 002 243591-DO-AVO-002-V01_251008 MEJORA DE LA ACCESIBILIDAD DE LA ESTACIÓN DE SANTA PERPÈTUA DE LA MOGODA – LA FLORIDA, BARCELONA Fecha Lugar Hora de inicio Hora de Finalización 08/10/25 Estación de Santa perpètua de la Mogoda - La Florida. 10:45 13:00 Principales asistentes: Nombre Empresa Cargo E-mail Teléfono Firma opcional FUNCIÓN: DC (director de contrato) DO (dirección de obra) DEO (dirección de ejecución) UTE (contratista) CSS (coordinador de seguridad) OBJETO Vi

--- Chunk 2 ---
tratista) CSS (coordinador de seguridad) OBJETO Visita Semanal de obra DOCUMENTACIÓN ENTREGADA O A ENTREGAR DOCUMENTACIÓN DE CONSULTA Proyecto Constructivo de Mejora de la accesibilidad de la estación de Santa Perpetua de la Mogoda – La Florida, Barcelona. ASUNTOS TRATADOS, ACUERDOS ALCANZADOS, COMUNICACIONES, INSTRUCCIONES: Nº Fecha DESCRIPCIÓN DEL ASUNTO: 1. AVANCE DE LA OBRA: 1.1. DO-DEO Visita semanal de obra 08/10/25 Con fecha de

In [45]:
# Dataset estructurado
data = []

for i, chunk in enumerate(chunks):
    data.append({
        "doc_id": "acta_obra_1",
        "chunk_id": i,
        "text": chunk
    })



In [46]:
print(len(data))
# 16 chunks

16


In [47]:
# Para ver el Data
for item in data:
    print(item)
    print("\n---\n")

{'doc_id': 'acta_obra_1', 'chunk_id': 0, 'text': 'ACTA DE VISITA DE OBRA Nº 002 243591-DO-AVO-002-V01_251008 MEJORA DE LA ACCESIBILIDAD DE LA ESTACIÓN DE SANTA PERPÈTUA DE LA MOGODA – LA FLORIDA, BARCELONA Fecha Lugar Hora de inicio Hora de Finalización 08/10/25 Estación de Santa perpètua de la Mogoda - La Florida. 10:45 13:00 Principales asistentes: Nombre Empresa Cargo E-mail Teléfono Firma opcional FUNCIÓN: DC (director de contrato) DO (dirección de obra) DEO (dirección de ejecución) UTE (contratista) CSS (coordinador de seguridad) OBJETO Vi'}

---

{'doc_id': 'acta_obra_1', 'chunk_id': 1, 'text': 'tratista) CSS (coordinador de seguridad) OBJETO Visita Semanal de obra DOCUMENTACIÓN ENTREGADA O A ENTREGAR DOCUMENTACIÓN DE CONSULTA Proyecto Constructivo de Mejora de la accesibilidad de la estación de Santa Perpetua de la Mogoda – La Florida, Barcelona. ASUNTOS TRATADOS, ACUERDOS ALCANZADOS, COMUNICACIONES, INSTRUCCIONES: Nº Fecha DESCRIPCIÓN DEL ASUNTO: 1. AVANCE DE LA OBRA: 1.1. DO-D

In [48]:
#  Guarda el resultado
import json

with open("chunks_acta.json", "w") as f:
    json.dump(data, f, indent=4)

In [49]:
# Descargar
from google.colab import files
files.download("chunks_acta.json")


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Segundo paso. Crear embeddings con BERT

In [50]:
# Importar BERT
from transformers import AutoTokenizer, AutoModel
import torch


In [51]:
# Descargar el modelo
model_name = "BSC-LT/MrBERT-es"

tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModel.from_pretrained(model_name)


/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:93: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


config.json: 0.00B [00:00, ?B/s]

tokenizer_config.json: 0.00B [00:00, ?B/s]

tokenizer.json:   0%|          | 0.00/6.83M [00:00<?, ?B/s]

tokenizer.model:   0%|          | 0.00/1.07M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/751 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/601M [00:00<?, ?B/s]

Loading weights:   0%|          | 0/134 [00:00<?, ?it/s]

ModernBertModel LOAD REPORT from: BSC-LT/MrBERT-es
Key               | Status     |  | 
------------------+------------+--+-
head.norm.weight  | UNEXPECTED |  | 
decoder.bias      | UNEXPECTED |  | 
head.dense.weight | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.


In [53]:
# Aplicar el modelo a los chunks
embeddings = []

for item in data:
    inputs = tokenizer(item["text"], return_tensors="pt", truncation=True, padding=True)

    with torch.no_grad():
        outputs = model(**inputs)

    embedding = outputs.last_hidden_state[:, 0, :].squeeze().tolist()

    embeddings.append({
        "doc_id": item["doc_id"],
        "chunk_id": item["chunk_id"],
        "embedding": embedding,
        "text": item["text"]
    })

In [58]:
# Define your search query here
search_query = "instalaciones de sistemas y megafonía" # @param {type:"string"}

In [59]:
from transformers import AutoTokenizer, AutoModel
import torch
import faiss
import numpy as np

# Ensure the tokenizer and model are loaded (if not already in scope)
model_name = "BSC-LT/MrBERT-es"
tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModel.from_pretrained(model_name)

# Embed the search query
query_inputs = tokenizer(search_query, return_tensors="pt", truncation=True, padding=True)

with torch.no_grad():
    query_outputs = model(**query_inputs)

query_embedding = query_outputs.last_hidden_state[:, 0, :].squeeze().numpy().astype('float32')

# Reshape query embedding for FAISS (FAISS expects 2D array)
query_embedding = query_embedding.reshape(1, -1)

# Perform similarity search
k = 5 # Number of nearest neighbors to retrieve
distances, indices = index.search(query_embedding, k)

print(f"Top {k} most similar chunks for query: '{search_query}'")
for i, idx in enumerate(indices[0]):
    print(f"\n--- Rank {i+1} (Distance: {distances[0][i]:.4f}) ---")
    print(f"Chunk ID: {embeddings[idx]['chunk_id']}")
    print(f"Text: {embeddings[idx]['text']}")

Loading weights:   0%|          | 0/134 [00:00<?, ?it/s]

ModernBertModel LOAD REPORT from: BSC-LT/MrBERT-es
Key               | Status     |  | 
------------------+------------+--+-
head.norm.weight  | UNEXPECTED |  | 
decoder.bias      | UNEXPECTED |  | 
head.dense.weight | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.


Top 5 most similar chunks for query: 'instalaciones de sistemas y megafonía'

--- Rank 1 (Distance: 1264.9529) ---
Chunk ID: 6
Text: . 2.2. Planning semanal / quincenal 08/10/25 Trabajos previstos para las próximas 2 semanas: Colocación barrera New Jersey en el límite entre viviendas y pie de talud. Levantamiento / informe fotográfico del estado actual de viviendas y entorno. Desbroce y o rebajes para la adecuación de la plataforma de acceso de máquinas. Terraplenado y compactación de la plataforma de accesos para la maquinaria de obra. Terminar la obra de conexión de pluviales. Traslado de la posición de la farola en el acce

--- Rank 2 (Distance: 1352.2341) ---
Chunk ID: 3
Text: 5 Implantación de obra: Montaje de casetas e instalaciones de obra y su señalización. Obra de conexión de saneamiento: Excavación, colocación de tubo, relleno compacto. 1.4. Breve descripción del avance 08/10/25 Se constata con la visita de obra que la contrata ha completado los trabajos iniciales de implanta

In [56]:
# Install FAISS
!pip install faiss-cpu

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 23.8/23.8 MB 65.9 MB/s eta 0:00:00


In [57]:
import faiss
import numpy as np

# Extract just the embedding vectors from the embeddings list
embedding_vectors = [item['embedding'] for item in embeddings]

# Convert the list of embeddings to a NumPy array
# Ensure all embedding vectors have the same dimension
dimension = len(embedding_vectors[0])
embedding_array = np.array(embedding_vectors).astype('float32')

# Create a FAISS index (IndexFlatL2 is a simple Euclidean distance index)
index = faiss.IndexFlatL2(dimension)

# Add the embeddings to the index
index.add(embedding_array)

print(f"FAISS index created with {index.ntotal} embeddings.")

FAISS index created with 16 embeddings.


The embeddings have been successfully stored in a FAISS index. You can now perform similarity searches on this index.

## Executive Summary of Document Analysis

This analysis involved extracting, processing, and indexing text from a provided DOCX document for efficient information retrieval. The key steps and findings are as follows:

### 1. Document Extraction and Cleaning
- The document **"243591-DO-AVO-02-V01-251008.docx"** was successfully loaded and its text content extracted.
- A cleaning process was applied to remove excess whitespace, resulting in a refined text of **7100 characters** (reduced from 7434 characters).

### 2. Text Chunking
- The cleaned text was divided into **16 overlapping chunks** (each with a size of 500 characters and an overlap of 50 characters) to facilitate granular analysis.
- These chunks were structured with `doc_id` and `chunk_id` for easy identification.

### 3. Embeddings Generation
- The **`BSC-LT/MrBERT-es`** transformer model was used to generate dense vector embeddings for each text chunk. These embeddings capture the semantic meaning of each chunk.

### 4. Vector Database (FAISS) Setup
- A **FAISS `IndexFlatL2` index** was created and populated with the 16 generated embeddings. This enables rapid similarity searches against the document's content.

### 5. Similarity Search Results
- A query, **"instalaciones de sistemas y megafonía"**, was embedded and used to search the FAISS index for the most relevant text segments. The top 5 results were:
    - **Rank 1 (Chunk ID 6):** Mentions **`Desmontaje acopio y o retirada de instalaciones anti-intrusión, megafonía y cartelería de señalización informativa de andén.`**
    - **Rank 2 (Chunk ID 3):** Discusses **`Implantación de obra: Montaje de casetas e instalaciones de obra y su señalización.`**
    - **Rank 3 (Chunk ID 11):** Refers to **`Conexión de acometida pluvial`**, which is less directly related but might share contextual terms.
    - **Rank 4 (Chunk ID 7):** Details **`Desmontaje acopio y o retirada de instalaciones anti-intrusión, megafonía y cartelería de señalización informativa de andén.`**
    - **Rank 5 (Chunk ID 2):** Mentions **`Implantación de obra: Montaje de casetas e instalaciones de obra`**, similar to Rank 2.

These results indicate that the system successfully identified the most semantically relevant sections of the document related to the query, particularly those explicitly mentioning 'megafonía' and 'instalaciones'.

In [55]:
display(embeddings[:2])

[{'doc_id': 'acta_obra_1',
  'chunk_id': 0,
  'embedding': [0.09689431637525558,
   1.3245662450790405,
   0.4162359833717346,
   0.031629402190446854,
   -0.04608100652694702,
   -0.19460700452327728,
   -1.2085381746292114,
   1.5572165250778198,
   -0.3478354215621948,
   -0.2871897518634796,
   0.1440146267414093,
   0.5179569721221924,
   0.698835551738739,
   -0.11634502559900284,
   -0.1094353049993515,
   -0.2731861174106598,
   1.7873725891113281,
   0.04151545837521553,
   0.5998106598854065,
   -1.2468606233596802,
   0.23898327350616455,
   1.1630258560180664,
   0.17326395213603973,
   -0.29139435291290283,
   -0.7338528633117676,
   1.229293704032898,
   -0.6302154064178467,
   -0.890778124332428,
   -0.7843397259712219,
   -1.610983967781067,
   0.6175293326377869,
   -0.6959841251373291,
   0.642021656036377,
   -0.4758875072002411,
   -43.235191345214844,
   -0.3789845407009125,
   -1.1290696859359741,
   1.9022183418273926,
   -0.30532917380332947,
   0.97077852487564

In [62]:
import json

with open('chunks_acta.json', 'r') as f:
    loaded_data = json.load(f)

# You can now use 'loaded_data' for further processing.
# For example, to verify its content:
print(f"Loaded {len(loaded_data)} chunks from 'chunks_acta.json'")
# display(loaded_data[:2]) # Display first two items to check

Loaded 16 chunks from 'chunks_acta.json'
